File: algo/76.py
from collections import Counter

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        missing = len(t)
        needs = Counter(t)
        start, end, lp = 0, 0, 0

        for rp, c in enumerate(s, 1):
            missing -= needs[c] > 0
            needs[c] -= 1

            if missing == 0:
                while lp < rp and needs[s[lp]] < 0:
                    needs[s[lp]] += 1
                    lp += 1

                if not end or rp - lp < end - start:
                    start, end = lp, rp
        
        return s[start:end]

s = "ADOBECODEBANC"
t = "ABC"
print(Solution().minWindow(s, t))

s = "a"
t = "a"
print(Solution().minWindow(s, t))

s = "a"
t = "aa"
print(Solution().minWindow(s, t))

File: old/algo/76.kt
class Solution {
    fun minWindow(s: String, t: String): String {
        val needs = HashMap<Char, Int>()
        for (ch in t) {
            needs[ch] = needs.getOrDefault(ch, 0) + 1
        }
        var missing = t.length
        var (start, end) = 0 to 0

        var lp = 0
        for ((rp, ch) in s.withIndex().map { (key, value) -> (key + 1) to value}) {
            if ((needs[ch] ?: 0) > 0) missing--
            needs[ch] = (needs[ch] ?: 0) - 1

            if (missing == 0) {
                while (lp < rp && (needs[s[lp]] ?: 0) < 0) {
                    needs[s[lp]] = (needs[s[lp]] ?: 0) + 1
                    lp++
                }

                if (end == 0 || rp - lp < end - start) {
                    start = lp
                    end = rp
                }
            }
        }
        return s.substring(start, end)
    }
}

//val s = "ADOBECODEBANC"
//val t = "ABC"

//val s = "a"
//val t = "a"

val s = "a"
val t = "aa"
println(Solution().minWindow(s, t))